include/solvers/navier_stokes_scratch_data.h

```diff
@@ -728,34 +707,60 @@ class NavierStokesScratchData : public PhysicsScratchDataBase
         for (unsigned int j = 0; j < number_of_particles; ++j)
           cell_void_fraction[j] = cell_void_fraction_bulk;
       }
+  }
 
-    if (number_of_particles == 0)
-      return;
-
-
-    // Calculate the average particle velocity within
-    // the cell
-    average_particle_velocity = average_particle_velocity / number_of_particles;
+  /** @brief Creates an object of type Quadrature<dim> that contains the
+   * particle reference location. This object can be used to interpolate any
+   * function known at the dofs at the location of the particles.
+   */
 
+  Quadrature<dim>
+  calculate_particle_location_quadrature_points()
+  {
     // Create local vector that will be used to spawn an in-situ quadrature to
-    // interpolate at the location of the particles
-    std::vector<Point<dim>> particle_reference_location;
+    // interpolate at the locations of the particles
+    std::vector<Point<dim>> particle_reference_location(number_of_particles);
     std::vector<double>     particle_weights(number_of_particles, 1);
+    unsigned int            particle_i = 0;
 
     // Loop over particles in cell and cache their reference location
     for (auto &particle : pic)
       {
         // Store particle positions and weights
         // Reference location of the particle
-        particle_reference_location.push_back(
-          particle.get_reference_location());
+        particle_reference_location[particle_i] =
+          particle.get_reference_location();
+        particle_i++;
       }
 
-    // Create a quadrature for the Navier-Stokes equations that is based on the
+    // Return a quadrature for the Navier-Stokes equations that is based on the
     // particle reference location
-    Quadrature<dim> q_local(particle_reference_location, particle_weights);
-    FEValues<dim>   fe_values_local_particles(this->fe_values.get_fe(),
-                                            q_local,
+    return Quadrature<dim>(particle_reference_location, particle_weights);
+  }
+
+  /** @brief Interpolates the velocity and pressure of the fluid, as well as the
+   * pressure gradient, and the laplacian, curl and gradient of the velocity, at
+   * the locations of the particles.
+   *
+   * @param previous_solution The solution (velocity and pressure) at the previous time step,
+   * which is used to interpolate the velocity and pressure at the particles
+   * locations.
+   *
+   * @param q_particles Quadrature type object that contains the location of the particles
+   * relative to the cell's frame of reference.
+   *
+   * @param dof_handler The DoFHandler of the Navier-Stokes problem
+   */
+
+  template <typename VectorType>
+  void
+  calculate_fluid_fields_at_particle_location(
+    const DoFHandler<dim> &dof_handler,
+    const Quadrature<dim> &q_particles,
+    const VectorType      &previous_solution)
```

Comment:
is it really a previous solution or just a solution? I feel the function is agnostic of the solution being previous or not. I would just call the Vector solution.